ui/desktop/electron-app/src/session/index.js

```diff
@@ -0,0 +1,118 @@
+const { spawnAsyncJSONPromise } = require('../spawn-promise.js');
+
+// FIXME create structure 'util' folder for spawn-promise
+
+/**
+ * Super paranoid shell quote/escape and validation.  Input must be base62.
+ * @param {string} str
+ */
+const escapeAndValidateBase62 = (str) => {
+  const candidate = str.toString();
+  if (candidate.match(/^[A-Za-z0-9_]*$/)) return candidate;
+  throw new Error(`
+    Could not invoke command:
+    input contained unsafe characters.
+  `);
+};
+
+class Session {
+  #id;
+  #pid;
+  #addr;
+  #token;
+  #hostId;
+  #process;
+  #targetId;
+  #proxyDetails;
+
+  /**
+   * Initialize a session to a controller address
+   * using target details.
+   * @param {string} addr
+   * @param {string} targetId
+   * @param {string} token
+   * @param {string} hostId
+   */
+  constructor(addr, targetId, token, hostId) {
+    this.#addr = addr;
+    this.#targetId = targetId;
+    this.#token = token;
+    this.#hostId = hostId;
+  }
+
+  /**
+   * Session id
+   * @return {string}
+   */
+  get id() {
+    return this.#id;
+  }
+
+  /**
+   * @return {boolean}
+   */
+  get isActive() {
+    return this.#process.connected;
+  }
+
+  /**
+   * Returns session proxy details.
+   * Only available after successful start of a session.
+   * @return {object}
+   */
+  get proxyDetails() {
+    return this.#proxyDetails;
+  }
+
+  /**
+   * Using cli, initialize a session to a target.
+   * Tracks local proxy details if successful.
+   */
+  start() {
+    const command = this.cliCommand();
+    return spawnAsyncJSONPromise(command).then((spawnedSession) => {
```

Comment:
You might return proxy details from the start promise to make it easier to get these.